datafusion/functions/src/regex/regexpcount.rs

```diff
@@ -0,0 +1,561 @@
+}
+
+impl Default for RegexpCountFunc {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl RegexpCountFunc {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::one_of(
+                vec![
+                    Exact(vec![Utf8, Utf8]),
+                    Exact(vec![Utf8, Utf8, Int64]),
+                    Exact(vec![Utf8, Utf8, Int64, Utf8]),
+                    Exact(vec![Utf8, Utf8, Int64, LargeUtf8]),
+                    Exact(vec![LargeUtf8, LargeUtf8]),
+                    Exact(vec![LargeUtf8, LargeUtf8, Int64]),
+                    Exact(vec![LargeUtf8, LargeUtf8, Int64, Utf8]),
+                    Exact(vec![LargeUtf8, LargeUtf8, Int64, LargeUtf8]),
+                ],
+                Volatility::Immutable,
+            ),
+        }
+    }
+}
+
+impl ScalarUDFImpl for RegexpCountFunc {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "regexp_count"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        use DataType::*;
+
+        Ok(match &arg_types[0] {
+            Null => Null,
+            _ => Int64,
+        })
+    }
+
+    fn invoke(&self, args: &[datafusion_expr::ColumnarValue]) -> Result<ColumnarValue> {
+        let len = args
+            .iter()
+            .fold(Option::<usize>::None, |acc, arg| match arg {
+                ColumnarValue::Scalar(_) => acc,
+                ColumnarValue::Array(a) => Some(a.len()),
+            });
+
+        let is_scalar = len.is_none();
+        let inferred_length = len.unwrap_or(1);
+        let args = args
+            .iter()
+            .map(|arg| arg.clone().into_array(inferred_length))
+            .collect::<Result<Vec<_>>>()?;
+
+        let result = regexp_count_func(&args);
+        if is_scalar {
+            // If all inputs are scalar, keeps output as scalar
+            let result = result.and_then(|arr| ScalarValue::try_from_array(&arr, 0));
+            result.map(ColumnarValue::Scalar)
+        } else {
+            result.map(ColumnarValue::Array)
+        }
+    }
+}
+
+fn regexp_count_func(args: &[ArrayRef]) -> Result<ArrayRef> {
+    match args[0].data_type() {
+        DataType::Utf8 => regexp_count::<i32>(args),
+        DataType::LargeUtf8 => regexp_count::<i64>(args),
+        other => {
+            internal_err!("Unsupported data type {other:?} for function regexp_count")
+        }
+    }
+}
+
+pub fn regexp_count<O: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+    let arg_len = args.len();
+
+    match arg_len {
+        2..=4 => {
+            let values = as_generic_string_array::<O>(&args[0])?;
+            let regex = as_generic_string_array::<O>(&args[1])?;
+            let regex_datum: &dyn Datum = if regex.len() != 1 {
+                regex
+            } else {
+                &Scalar::new(regex)
```

Comment:
The branching feature is only available starting rust 1.80.0. Otherwise, we need to define branches for every combination, e.g.,

```

            if regex.len() > 1 && start.len() > 1 && flags.len() > 1 {
                regexp_array_count(&values, regex, Some(&start), Some(&flags))
            } else if regex.len() > 1 && start.len() > 1 {
                let regex_scalar = Scalar::new(regex);
                regexp_array_count(&values, &regex_scalar, Some(&start), Some(&flags))
            } else if regex.len() > 1 && flags.len() > 1 {
                let regex_scalar = Scalar::new(regex);
                regexp_array_count(&values, &regex_scalar, Some(&start), Some(&flags))
            } else if start.len() > 1 && flags.len() > 1 {
                let start_scalar = Scalar::new(start);
                regexp_array_count(&values, regex, Some(&start_scalar), Some(&flags))
            } else if regex.len() > 1 {
                let regex_scalar = Scalar::new(regex);
                regexp_array_count(&values, &regex_scalar, Some(&start), Some(&flags))
            } else if start.len() > 1 {
                let start_scalar = Scalar::new(start);
                regexp_array_count(&values, regex, Some(&start_scalar), Some(&flags))
            } else if flags.len() > 1 {
                let flags_scalar = Scalar::new(flags);
                regexp_array_count(&values, regex, Some(&start), Some(&flags_scalar))
            } else {
                let regex_scalar = Scalar::new(regex);
                let start_scalar = Scalar::new(start);
                let flags_scalar = Scalar::new(flags);
                regexp_array_count(&values, &regex_scalar, Some(&start_scalar), Some(&flags_scalar))
            }
```

Reply:
I am not Rust expert. Looking into their changelogs, this should named `Propagate temporary lifetime extension into if and match. ` stablized in Rust 1.79.

For your interests, https://github.com/rust-lang/rust/pull/121346/

Reply:
MSRV for DataFusion is currently [1,76](https://github.com/apache/datafusion/blob/945902dd5d440bdc360cab60ef31cd0c3bceec41/Cargo.toml#L60C1-L60C22)

Reply:
This is no longer required. Close.